src/map/clif.c

```diff
@@ -20186,6 +20186,162 @@ void clif_skill_scale(struct block_list *bl, int src_id, int x, int y, uint16 sk
+		clif->attendancedb_libconfig_sub(it, i, config_filename);
+	}
+
+	libconfig->destroy(&attendance_conf);
+	ShowStatus("Done reading '"CL_WHITE"%d"CL_RESET"' entries in '"CL_WHITE"%s"CL_RESET"'.\n", i, config_filename);
+	return true;
+}
+
+bool clif_attendancedb_libconfig_sub(struct config_setting_t *it, int n, const char *source) {
+	struct config_setting_t *t = NULL;
+	struct item_data *data = NULL;
+	int i32 = 0;
+	const char *str = NULL;
+
+	nullpo_ret(source);
+	Assert_ret(n >= 0 && n < MAX_ATTENDANCE_DAYS);
+
+	if (!libconfig->setting_lookup_string(it, "ItemID", &str) || !*str) {
+		ShowWarning("clif_attendancedb_libconfig_sub: Invalid entrie in \"%s\", entry #%d, skipping.\n", source, n);
+		return false;
+	}
+	if (!(data = itemdb->name2id(str))) {
+		ShowWarning("clif_attendancedb_libconfig_sub: unknown item \"%s\", entry #%d, skipping.\n", str, n);
+		return false;
+	}
+	clif->attendance_data.nameid[n] = (uint16)data->nameid;
+
+	if ((t = libconfig->setting_get_member(it, "Amount")) == NULL || (i32 = libconfig->setting_get_int(t)) < 1) {
+		ShowWarning("clif_attendancedb_libconfig_sub: Invalid or missing amount in \"%s\", entry #%d, skipping.\n", source, n);
+		return false;
+	}
+	clif->attendance_data.qty[n] = i32;
+
+	return true;
+}
+
+bool clif_attendance_timediff(struct map_session_data *sd) {
+
+	nullpo_retr(false, sd);
+
+	int64 timediff = (time(NULL) / (60 * 60 * 24)) - (sd->status.attendance_timer / (60 * 60 * 24));
+
+	if (timediff <= 0)
+		return false;
+	return true;
+}
+
+void clif_parse_open_ui_request(int fd, struct map_session_data *sd) __attribute__((nonnull(2)));
+void clif_parse_open_ui_request(int fd, struct map_session_data *sd)
+{
+
+	if (battle_config.feature_enable_attendance_system != 1)
+		return;
+
+	const struct PACKET_CZ_OPEN_UI *p = RFIFOP(fd, 0);
+
+	clif->open_ui(sd, p->UIType);
+}
+
+void clif_open_ui(struct map_session_data *sd, int8 UIType) {
+#if defined(PACKETVER_RE) && PACKETVER >= 20180307
+
+	nullpo_retv(sd);
+
+	struct PACKET_ZC_OPEN_UI p;
+	int claimed = 0;
+
+	p.PacketType = 0xAE2;
+	switch (UIType) {
+	case 5: // receive 5 for ATTENDANCE_UI
+		if (clif->attendance_timediff(sd) != true)
+			++claimed;
+		else if (sd->status.attendance_count >= MAX_ATTENDANCE_DAYS)
+			sd->status.attendance_count = 0;
+		p.UIType = ATTENDANCE_UI;
+		p.data = sd->status.attendance_count * 10 + claimed;
+		break;
+	default:
+		ShowWarning("clif_open_ui: Requested UI (%d) is not implemented yet.\n", UIType);
+		return;
+	}
+
+	clif->send(&p, sizeof(p), &sd->bl, SELF);
+#else
+	ShowWarning("Attendance System available only for PACKETVER_RE && (PACKETVER >= 20180307).\n");
+#endif
+}
+
+void clif_parse_attendance_reward_request(int fd, struct map_session_data *sd) __attribute__((nonnull(2)));
+void clif_parse_attendance_reward_request(int fd, struct map_session_data *sd)
+{
+#if defined(PACKETVER_RE) && PACKETVER >= 20180307
+
+	if (battle_config.feature_enable_attendance_system != 1)
+		return;
+
+	nullpo_retv(sd);
```

Comment:
here sd check useless, because __attribute__((nonnull(2)) said it must be non null